File: Data-Structures/Trees/0144_binary_tree_preorder.c
// Practice for: 0144_binary_tree_preorder.py
// Bridge logic from Python to C pointers here.

#include <stdio.h>
#include <stdlib.h>

struct TreeNode {
    int val;
    struct TreeNode *left;
    struct TreeNode *right;
};

int* preorderTraversal(struct TreeNode *root, int *returnSize){
    //LIFO, always look at left child first 
    int* res = (int*)malloc(100 * sizeof(int));
    *returnSize = 0;

    if (root == NULL){
        return res;
    }
    
    struct TreeNode *stack[100];
    int top = 0;

    stack[top++] = root;

    while (top > 0){
        //pop top element 
        struct TreeNode *node = stack[--top];
        res[(*returnSize)++] = node -> val;

        if (node->right){
            stack[top++] = node->right;
        }

         if (node->left){
            stack[top++] = node->right;
        }
    }
    return res;
}

File: Data-Structures/Trees/0144_binary_tree_preorder.py
#trees

from typing import Optional

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
class Solution:
    def preorderTraversal(self, root: Optional[TreeNode]) -> List[int]:
        #LIFO, process left child first always
        res= []

        if not root: 
            return res
        
        #initalise stack 
        st = [root]

        #pop current element, then add it to res
        while st:
            node = st.pop()
            res.append(node.val)

            if node.right:
                st.append(node.right)
            
            if node.left:
                st.append(node.left)
        return res